docs(adr): Public API (#506)
celestiaorg/celestia-node: docs/adr/adr-009-public-api.md

	GetShare(ctx context.Context, dah *Root, row, col int) (Share, error)
	// GetSharesByNamespace returns all shares of the given nID from the given data Root.
    GetSharesByNamespace(ctx context.Context, root *Root, nID namespace.ID) ([]Share, error)
    // SharesAvailable subjectively validates if Shares committed to the given data Root are 
	// available on the network. 
    SharesAvailable(ctx context.Context, root *Root) error
    // ProbabilityOfAvailability calculates the probability of the data square
    // being available based on the number of samples collected.
    ProbabilityOfAvailability() float64
}
```

##### P2P
```go
type P2PModule interface {
    Info
    Peers
    PeerInfo

    Connect
    Disconnect
    Connectedness
    NATStatus

    BlockAdd
    BlockRm
    BlockList

    MutualAdd
    MutualRm
    MutualList

    BandwidthStats
    BandwidthPeer
    BandwidthProtocol

    ResourceState
    ResourceLimit
    ResourceSetLimit

    PubSubPeers
    PubSubScore

    BitSwapStats
    BitSwapLedger

    DHTFindPeer
}
```

##### Wallet
```go
type WalletModule interface {
    New
    Del
    Has
    List
    Sign
    Export
    Import
    // TODO: Mnemonics stuff?
}

> ```

Reviewer: I want to move the wallet module to _nice to have_ because we have the `cel-key` utility which is good ~enough~ for now.

Reviewer 2: This separation is good.

Also, "key utility" is IMHO way better name than "WalletModule" (because wallets typically shows your bags, let you see history, do transfers, etc).

### NodeModule
```go
type NodeModule interface {
    Type
    Version

    LogList
    LogSet

	Start(ctx context.Context) error
	Stop(ctx context.Context) error

    // TODO: Auth
}
```

##### DAS

> // TODO: Should we rebrand DASer to ShareSync?

Reviewer: IMO no. We need to retain the key word "sample" bc sharesync implies a full sync of all shares

### Implementation

The tracking issue can be found [here](https://github.com/celestiaorg/celestia-node/issues/944).
It provides a more detailed step-by-step process for how the below described design will be implemented.

### Details
#### Services Deprecation
The initial step is to deprecate services in favor of modules. Ex. `HeaderService` -> `HeaderModule`.

* We're organically moving towards the direction of modularized libraries. That is, our `share`, `header` and `state`
packages are getting shaped as independent modules which now lack their own API definition. 
* Consistency. Semantically, modules are closer to what high level Celestia project goals and slogans.
* Disassociate centralization. Services have always been associated with centralized infrastructure with older 
monolithic services and newer distributed microservice architectures. 

#### Modules Definition

##### Header
```go
type HeaderModule interface {
	// Head returns the node's local head (tip of the chain of the header store).
	Head(ctx context.Context) (*header.ExtendedHeader, error)
	// Get returns the header of the given hash from the node's header store.

> 	Get(ctx context.Context, hash tmbytes.HexBytes) (*header.ExtendedHeader, error)

Author: We could. If we want to go into the direction of general purpose Header lib, we could also consider a general purpose hash definition

> 	Get(ctx context.Context, hash tmbytes.HexBytes) (*header.ExtendedHeader, error)

Reviewer: We could use plain []byte but `HexBytes` provides some nice convenience methods (e.g. `String()`, `MarshalJSON()`) that could be useful. We can also alias it? Just something to consider.

    Import
    // TODO: Mnemonics stuff?
}
```

### NodeModule
```go
type NodeModule interface {
    Type
    Version

    LogList
    LogLevelSet

	Start(ctx context.Context) error
	Stop(ctx context.Context) error

    // TODO: Auth
}
```

##### DAS
// TODO: Should we rebrand DASer to ShareSync?
```go
type DASModule interface {
	// State reports the current state of the DASer.
    State() das.State
}
```

##### State
```go
type StateModule interface {
    // Balance retrieves the Celestia coin balance for the node's account/signer
    // and verifies it against the corresponding block's AppHash.
    Balance(ctx context.Context) (*state.Balance, error)
    // BalanceForAddress retrieves the Celestia coin balance for the given address and verifies
    // the returned balance against the corresponding block's AppHash.
    BalanceForAddress(ctx context.Context, addr state.Address) (*state.Balance, error)
    // SubmitTx submits the given transaction/message to the Celestia network and blocks until 
	// the tx is included in a block.
    SubmitTx(ctx context.Context, tx state.Tx) (*state.TxResponse, error)
    // SubmitPayForData builds, signs and submits a PayForData transaction.

>     SubmitPayForData(ctx context.Context, nID namespace.ID, data []byte, gasLimit uint64) (*state.TxResponse, error)

Reviewer 2: I think it will be good to let users use different key management. 
This will require to expose a method to query account numbers or build entire PFD without signing.

Reviewer: I guess we can expose an additional `BuildPayForData` and they're responsible for signing it and submitting it via `SubmitTx`

Reviewer: @tzdybal is this separation something you need before mainnet? Otherwise I'll stick it in "nice to have post mainnet"

Reviewer 2: It's definitely not critical before mainnet, not a blocker.

Author: @renaynay, we should introduce like a `SmthOptions` structure and put it as the last param asap. Then we can add more optional fields without breaking the interface. 

The question I still have here is: should we make such optional fields structure per module or per method? Per module seems fine if modules are granular enough.

Author: Another alternative is to add methods for each optional use case, but I don't really like this, as it bloats the interface.

    // Transfer sends the given amount of coins from default wallet of the node to the given account address.

>     Transfer(ctx context.Context, to types.Address, amount types.Int, gasLimit uint64) (*state.TxResponse, error)

Reviewer 2: `Transfer` is related to `State` but it's also feels like a `wallet` operation.

Reviewer: We probably won't have a wallet module for mainnet.

Reviewer: We will have the key utility instead 👍🏻

    GetByHeight(ctx context.Context, height uint64) (*header.ExtendedHeader, error)
    // GetRangeByHeight returns the given range [from:to) of ExtendedHeaders.
    GetRangeByHeight(ctx context.Context, from, to uint64) ([]*ExtendedHeader, error)
    // Subscribe creates long-living Subscription for validated ExtendedHeaders.
    // Multiple Subscriptions can be created.
    Subscribe() (Subscription, error)    
	// SyncState returns the current state of the header Syncer. 
	SyncState() sync.State
	// SyncWait blocks until the header Syncer is synced to network head. 
    SyncWait(ctx context.Context) error
	// SyncHead provides the Syncer's view of the current network head.
	SyncHead(ctx context.Context) (*header.ExtendedHeader, error)
}
```

##### Shares
```go
type SharesModule interface {
	// GetShare returns the Share from the given data Root at the given row/col coordinates.
	GetShare(ctx context.Context, dah *Root, row, col int) (Share, error)
	// GetSharesByNamespace returns all shares of the given nID from the given data Root.
    GetSharesByNamespace(ctx context.Context, root *Root, nID namespace.ID) ([]Share, error)

>     // SharesAvailable subjectively validates if Shares committed to the given data Root are 
> 	// available on the network. 
>     SharesAvailable(ctx context.Context, root *Root) error

Reviewer 3: [optional][nit] prefixing this method with an `is` may make it clearer to readers that this method has boolean output.

Proposals: `isEveryShareAvailable`. The 99.9% is a placeholder

```suggestion
	// isEveryShareAvailable checks if the Shares committed to in the given data
	// Root are available on the network. Returns nil if higher than 99.9%
	// confidence the shares are available, otherwise returns an error.
	isEveryShareAvailable(ctx context.Context, root *Root) error
```

Ref: https://dev.to/michi/tips-on-naming-boolean-variables-cleaner-code-35ig

Update: on second thought the return value of this function is `error` so an `is` prefix may be misleading. If the `error` doesn't need to be propagated to callers, we could return `boolean` for clarity. If the error does need to be propagated, we could return `(boolean, error)`. I don't have strong opinions here.

    // ProbabilityOfAvailability calculates the probability of the data square
    // being available based on the number of samples collected.

>     ProbabilityOfAvailability() float64

Reviewer 3: [question] which data square is this method using if it doesn't accept any arguments?

Reviewer 4: It is the theoretical probability for any given data square - for example, it returns `1` on full nodes, and `1 - math.Pow(0.75, float64(DefaultSampleAmount))` on light nodes

    Import
    // TODO: Mnemonics stuff?
}
```

### NodeModule
```go
type NodeModule interface {
    Type
    Version

    LogList
    LogLevelSet

	Start(ctx context.Context) error
	Stop(ctx context.Context) error

    // TODO: Auth
}
```

##### DAS
// TODO: Should we rebrand DASer to ShareSync?
```go
type DASModule interface {
	// State reports the current state of the DASer.
    State() das.State
}
```

##### State
```go
type StateModule interface {
    // Balance retrieves the Celestia coin balance for the node's account/signer
    // and verifies it against the corresponding block's AppHash.
    Balance(ctx context.Context) (*state.Balance, error)
    // BalanceForAddress retrieves the Celestia coin balance for the given address and verifies
    // the returned balance against the corresponding block's AppHash.
    BalanceForAddress(ctx context.Context, addr state.Address) (*state.Balance, error)
    // SubmitTx submits the given transaction/message to the Celestia network and blocks until 
	// the tx is included in a block.
    SubmitTx(ctx context.Context, tx state.Tx) (*state.TxResponse, error)
    // SubmitPayForData builds, signs and submits a PayForData transaction.

>     SubmitPayForData(ctx context.Context, nID namespace.ID, data []byte, gasLim uint64) (*state.TxResponse, error)

Reviewer 3: [nit] for consistency with the argument in the the `Transfer()` function below

```suggestion
    SubmitPayForData(ctx context.Context, nID namespace.ID, data []byte, gasLimit uint64) (*state.TxResponse, error)
```

// SyncState returns the current state of the header Syncer. 
SyncState() sync.State
// SyncWait blocks until the header Syncer is synced to network head. 
SyncWait(ctx context.Context) error
// SyncHead provides the Syncer's view of the current network head.
SyncHead(ctx context.Context) (*header.ExtendedHeader, error)
}
```

##### Shares

```go
  type SharesModule interface {
    // GetShare returns the Share from the given data Root at the given row/col
    // coordinates.
    GetShare(ctx context.Context, dah *Root, row, col int) (Share, error)
    // GetSharesByNamespace returns all shares of the given nID from the given data
    // Root.
    GetSharesByNamespace(
      ctx context.Context,
      root *Root, 
      nID namespace.ID, 
    ) ([]Share, error)
    // SharesAvailable subjectively validates if Shares committed to the given data
    // Root are available on the network. 
    SharesAvailable(ctx context.Context, root *Root) error
    // ProbabilityOfAvailability calculates the probability of the data square
    // being available based on the number of samples collected.
    ProbabilityOfAvailability() float64
  }
```

##### P2P

```go

>   type P2PModule interface {

Reviewer: done

    Info
    Peers
    PeerInfo
    
    Connect
    Disconnect
    Connectedness
    NATStatus
    
    BlockAdd
    BlockRm
    BlockList
    
    MutualAdd
    MutualRm
    MutualList
    
    BandwidthStats
    BandwidthPeer
    BandwidthProtocol
    
    ResourceState
    ResourceLimit
    ResourceSetLimit
    
    PubSubPeers
    PubSubScore
    
    BitSwapStats
    BitSwapLedger
    
    DHTFindPeer
  }
```

### NodeModule

```go
  type NodeModule interface {
    Type
    Version
    
    LogLevelList(ctx context.Context) (map[string]string, error)
    LogLevelSet(ctx context.Context, name, level string) error
    
    Start(ctx context.Context) error
    Stop(ctx context.Context) error

    // NodeModule provides AuthModule 
    AuthModule
  }
```

#### AuthModule

```go
type AuthModule interface {
    SetModuleAuth(ctx context.Context, module string) error  // TODO

>     UnsetModuleAuth(ctx context.Context, module string) error // TODO

Author: This API limits to only one user who has auth/dawg/privileges, meaning the API doesn't allow us to have multiple authenticated users: CLI, Optimint, Astria. Or they should have shared access token, which is a bad security practice. 

We should be able to generate multiple access tokens through the API and potentially with another permissions dimension. So that the node operator can grant access to various apps.

The minimalistic example of such an API is https://github.com/filecoin-project/lotus/blob/master/api/api_common.go. Find Auth there.

}
```

#### DAS

> // TODO: Should we rebrand DASer to ShareSync?

Reviewer: done

```go
  type DASModule interface {
    // State reports the current state of the DASer.

>     State() das.SamplingState

Reviewer: done

  }
```

#### State

```go
  type StateModule interface {

>     // Balance retrieves the Celestia coin balance for the node's account/signer

Reviewer 5: ```suggestion
    // AccountAddress retrieves the address of the node's account/signer
    AccountAddress(ctx context.Context) (state.Address, error)
    // Balance retrieves the Celestia coin balance for the node's account/signer
```

Reviewer: done

    // address and verifies the returned balance against the corresponding
    // block's AppHash.
    BalanceForAddress(ctx context.Context, addr state.Address) (*state.Balance, error)
    // SubmitTx submits the given transaction/message to the Celestia network
    // and blocks until the tx is included in a block.
    SubmitTx(ctx context.Context, tx state.Tx) (*state.TxResponse, error)
    // SubmitPayForData builds, signs and submits a PayForData transaction.
    SubmitPayForData(
      ctx context.Context, 
      nID namespace.ID, 
      data []byte, 
      gasLimit uint64,
    ) (*state.TxResponse, error)
    // Transfer sends the given amount of coins from default wallet of the node 
    // to the given account address.
    Transfer(
      ctx context.Context, 
      to types.Address, 
      amount types.Int, 
      gasLimit uint64,
    ) (*state.TxResponse, error)

    // StateModule also provides StakingModule
    StakingModule
  }
```

Ideally all the state modules below should be implemented on top of only
StateModule, but there is no way we can have an abstract state requesting method,
yet.

##### Staking

```go
  type StakingModule interface {

>     Delegate

Reviewer: done

    Redelegate
    Unbond
    CancelUnbond
    
    QueryDelegation
    QueryRedelegation
    QueryUnbondingDelegation
  }
```

#### Fraud

```go
  type FraudModule interface {

>     Subscribe(type)

Author: We might allow Broadcasting of custom FPs eventually. This is a discussion for broader audience. For now let's just Subscribe

>     Subscribe(type)

Reviewer: Not broadcaster as that shouldn't be accessible IMO outside of the component that is in charge of constructing + broadcasting proof.

    Get(type)

>     List(type)

Author: Fraud Proofs

>     List(type)

Reviewer: check now

# ADR 009: Public API

## Authors

@Wondertan @renaynay

## Changelog

- 2022-03-08: initial version
- 2022-08-03: additional details

## Context

Celestia Node has been built for almost half a year with a bottom-up approach to
development. The core lower level components are built first and public API
around them is getting organically shaped. Now that the project is maturing and
its architecture is better defined, it's a good time to formally define a set of
modules provided by the node and their respective APIs.

## Alternative Approaches

### Node Type centric Design

Another discussed approach to defining an API could be an onion style, where
each new layer is a feature set of a broader node type. However, it turns out
that this approach does not seem to match the reality we have, where each node
type implements all the possible APIs but with the variable implementations
matching resource constraints of a type.

## Design

### Goals

- Ergonomic. Simple, idiomatic and self-explanatory.

> - Module centric(modular). The API is not monolithic and is segregated into

Reviewer 6: nit/non-blocking:
```suggestion
- Module centric (modular). The API is not monolithic and is segregated into
```

I see this a lot in other documentation and comments fwiw.

network layer (DA layer)` such that the node contains all functionality
necessary to post and retrieve messages from the DA layer.

This means that DA nodes will be able to query for / modify celestia state such
that the DA nodes are able to pay for posting their messages on the network.
The state-related API will be documented below in detail.

Furthermore, interaction between the celestia consensus network and the
celestia data availability network will be the responsibility of the **bridge**
node type. However, that interaction will not be exposed on a public level
(meaning a **bridge** node will not expose the same API as the
celestia-core node to which it is connected). A **bridge** node, for all intents
and purposes, will provide the same API as that of a **full** node (with a
stubbed-out DAS module as bridge nodes do not perform sampling).

### Details

#### Services Deprecation

The initial step is to deprecate services in favor of modules. Ex.
`HeaderService` -> `HeaderModule`.

- We're organically moving towards the direction of modularized libraries. That
is, our `share`, `header` and `state` packages are getting shaped as independent

> modules which now lack their own API definition.

Reviewer 6: Could be clearer: 
```suggestion
modules which currently lack their own API definition.
```
_now_ could also mean once the ADR is fully implemented.

Head(ctx context.Context) (*header.ExtendedHeader, error)
// Get returns the header of the given hash from the node's header store.
Get(ctx context.Context, hash tmbytes.HexBytes) (*header.ExtendedHeader, error)
// GetByHeight returns the header of the given height from the node's header store.
// If the header of the given height is not yet available, the request will hang
// until it becomes available in the node's header store.
GetByHeight(ctx context.Context, height uint64) (*header.ExtendedHeader, error)
// GetRangeByHeight returns the given range [from:to) of ExtendedHeaders.
GetRangeByHeight(ctx context.Context, from, to uint64) ([]*ExtendedHeader, error)
// Subscribe creates long-living Subscription for validated ExtendedHeaders.
// Multiple Subscriptions can be created.
Subscribe() (Subscription, error)
// SyncState returns the current state of the header Syncer. 
SyncState() sync.State
// SyncWait blocks until the header Syncer is synced to network head. 
SyncWait(ctx context.Context) error
// SyncHead provides the Syncer's view of the current network head.
SyncHead(ctx context.Context) (*header.ExtendedHeader, error)
}

> ```

Reviewer 6: This is not correctly/consistently formatted. Compare `SharesModule` code-block below for instance.

##### Shares

```go
  type SharesModule interface {
    // GetShare returns the Share from the given data Root at the given row/col
    // coordinates.

>     GetShare(ctx context.Context, dah *Root, row, col int) (Share, error)

Reviewer 7: nit: since it is `root` in all the other ones. 
```suggestion
    GetShare(ctx context.Context, root *Root, row, col int) (Share, error)
```

    // GetSharesByNamespace returns all shares of the given nID from the given data
    // Root.
    GetSharesByNamespace(
      ctx context.Context,
      root *Root, 
      nID namespace.ID, 
    ) ([]Share, error)
    // SharesAvailable subjectively validates if Shares committed to the given data
    // Root are available on the network. 
    SharesAvailable(ctx context.Context, root *Root) error
    // ProbabilityOfAvailability calculates the probability of the data square
    // being available based on the number of samples collected.
    ProbabilityOfAvailability() float64
  }
```

##### P2P

```go
  type P2PModule interface {
    // Info returns basic information about the node's p2p host/operations.
    Info() p2p.Info
    // Peers returns all peer IDs used across all inner stores.
    Peers() peer.IDSlice 
    // PeerInfo returns a small slice of information Peerstore has on the
    // given peer.
    PeerInfo(id peer.ID) peer.AddrInfo
   
    // Connect ensures there is a connection between this host and the peer with
    // given peer.ID. 

>     Connect(ctx context.Context, pi peer.AddrInfo) error

Reviewer 7: based on the comment
```suggestion
    Connect(ctx context.Context, pi peer.ID) error
```

    // ClosePeer closes the connection to a given peer. 
    ClosePeer(id peer.ID) error
    // Connectedness returns a state signaling connection capabilities.
    Connectedness(id peer.ID) network.Connectedness
    // NATStatus returns the current NAT status.
    NATStatus() network.Reachability
   
    // BlockPeer adds a peer to the set of blocked peers.
    BlockPeer(p peer.ID) error
    // UnblockPeer removes a peer from the set of blocked peers.
    UnblockPeer(p peer.ID) error
    // ListBlockedPeers returns a list of blocked peers.
    ListBlockedPeers() []peer.ID

    // MutualAdd adds a peer to the list of peers who have a bidirectional
    // peering agreement that they are protected from being trimmed, dropped
    // or negatively scored.
    MutualAdd(id peer.ID, tag string)

>     // MutualAdd removes a peer from the list of peers who have a bidirectional

Reviewer 6: ```suggestion
    // MutualRm removes a peer from the list of peers who have a bidirectional
```

    // peering agreement that they are protected from being trimmed, dropped
    // or negatively scored, returning a bool representing whether the given
    // peer is protected or not.

>     MutualRm(id peer.ID, tag string) bool

Reviewer 6: This is confusing: If I want to remove a peer and `MutualRm` returns `true`,  it means the peer was protected? Does `true` mean it does not get removed then?

```go
  type StateModule interface {
    // AccountAddress retrieves the address of the node's account/signer
    AccountAddress(ctx context.Context) (state.Address, error)
    // Balance retrieves the Celestia coin balance for the node's account/signer
    // and verifies it against the corresponding block's AppHash.
    Balance(ctx context.Context) (*state.Balance, error)
    // BalanceForAddress retrieves the Celestia coin balance for the given 
    // address and verifies the returned balance against the corresponding
    // block's AppHash.
    BalanceForAddress(ctx context.Context, addr state.Address) (*state.Balance, error)
    // SubmitTx submits the given transaction/message to the Celestia network
    // and blocks until the tx is included in a block.
    SubmitTx(ctx context.Context, tx state.Tx) (*state.TxResponse, error)
    // SubmitPayForData builds, signs and submits a PayForData transaction.
    SubmitPayForData(
      ctx context.Context, 
      nID namespace.ID, 
      data []byte, 
      gasLimit uint64,
    ) (*state.TxResponse, error)
    // Transfer sends the given amount of coins from default wallet of the node 
    // to the given account address.

>     Transfer(

Reviewer 7: [non-blocking]
Do we want a full balance transfer option?  Or have we discussed how someone would do that?

I assume the full cost of this method is amount + gasFee. So it's hard for users to completely empty a wallet. 
But if we can structure it to allow for amount being the full cost, and what is sent to the new address is amount - gasFee, then users can drain a wallet easily. 

Something a simple as a `subGass bool` argument or just a second method.

      ctx context.Context, 
      to types.Address, 
      amount types.Int, 
      gasLimit uint64,
    ) (*state.TxResponse, error)

    // StateModule also provides StakingModule
    StakingModule
  }
```

Ideally all the state modules below should be implemented on top of only
StateModule, but there is no way we can have an abstract state requesting method,
yet.

##### Staking

```go
  type StakingModule interface {
    // Delegate sends a user's liquid tokens to a validator for delegation.
    Delegate(
        ctx context.Context, 
        delAddr state.ValAddress, 
        amount state.Int, 
        gasLim uint64,
    ) (*state.TxResponse, error)
    // BeginRedelegate sends a user's delegated tokens to a new validator for redelegation.

>     BeginRedelegate(

Reviewer 7: nit: why no just `Redelegate`?

Reviewer: We keep the naming scheme of celestia-core

    ) (*types.QueryDelegationResponse, error)
    // QueryRedelegations retrieves the status of the redelegations between a
    // delegator and a validator.
    QueryRedelegations(
        ctx context.Context,
        srcValAddr,
        dstValAddr state.ValAddress,
    ) (*types.QueryRedelegationsResponse, error)
    // QueryUnbonding retrieves the unbonding status between a delegator and a validator.
    QueryUnbonding(
        ctx context.Context,
        valAddr state.ValAddress,
    ) (*types.QueryUnbondingDelegationResponse, error)

  }
```

#### Fraud

```go
  type FraudModule interface {
    // Subscribe subscribes to the given fraud proof type.
    Subscribe(proof fraud.ProofType) error
    // List lists all proof types to which the fraud module is currently
    // subscribed.
    List() []fraud.ProofType

    // Get returns any stored fraud proofs of the given type.
    Get(proof fraud.ProofType) ([]Proof, error)
  }
```

#### Metrics

```go
  type MetricsModule interface {
    // List shows all the registered meters.
    List(ctx) (string[], error)
    
    // Enable turns on the specific meter.
    Enable(string) error
    Disable(string) error
    

>     // ExportTo  sets the endpoint the metrics should be exported to

Reviewer 6: ```suggestion
    // ExportTo sets the endpoint the metrics should be exported to.
```

    // PeerInfo returns a small slice of information Peerstore has on the
    // given peer.
    PeerInfo(id peer.ID) peer.AddrInfo
   
    // Connect ensures there is a connection between this host and the peer with
    // given peer.ID. 
    Connect(ctx context.Context, pi peer.AddrInfo) error
    // ClosePeer closes the connection to a given peer. 
    ClosePeer(id peer.ID) error
    // Connectedness returns a state signaling connection capabilities.
    Connectedness(id peer.ID) network.Connectedness
    // NATStatus returns the current NAT status.
    NATStatus() network.Reachability
   
    // BlockPeer adds a peer to the set of blocked peers. // TODO should we wrap BlockPeer so that it 1. disconnects from peer, then 2. adds to blocklist?
    BlockPeer(p peer.ID) error
    // UnblockPeer removes a peer from the set of blocked peers.
    UnblockPeer(p peer.ID) error
    // ListBlockedPeers returns a list of blocked peers.
    ListBlockedPeers() []peer.ID

    // MutualAdd adds a peer to the list of peers who have a bidirectional
    // peering agreement that they are protected from being trimmed, dropped
    // or negatively scored.
    MutualAdd(id peer.ID, tag string)

>     // MutualAdd removes a peer from the list of peers who have a bidirectional

Reviewer: MutualRm

    // The metrics returned include all traffic sent / received for the peer, regardless of protocol.
    BandwidthForPeer(id peer.ID) Stats
    // BandwidthForProtocol returns a Stats struct with bandwidth metrics associated with the given protocol.ID.	
    BandwidthForProtocol(proto protocol.ID) Stats
   
    // ResourceState returns the state of the resource manager.
    ResourceState() rcmgr.ResourceManagerStat
    // ResourceLimit returns the limit of the resource scope. 
    ResourceLimit() rcmgr.Limit
    // ResourceSetLimit sets the limit of the resource scope.
    ResourceSetLimit(rcmgr.Limit)
   
    // PubSubPeers returns the peer IDs of the peers joined on
    // the given topic.
    PubSubPeers(topic string) ([]peer.ID, error)

    BitSwapStats // TODO ?
    BitSwapLedger // TODO ? 
    
    DHTFindPeer // TODO ?
  }
```

### NodeModule

```go
  type NodeModule interface {
    // Type returns the node type.
    Type() node.Type
    // Version returns information about the current binary build.
    Version() string
   
    // LogLevelList lists all the log levels per component.

>     LogLevelList(ctx context.Context) (map[string]string, error)

Reviewer: we cannot list log level per logger.

    // LogLevelSet sets the given component log level to the given level.
    LogLevelSet(ctx context.Context, name, level string) error
   
    // Start starts the node.
    Start(ctx context.Context) error
    // Stop stops the node.
    Stop(ctx context.Context) error

    // NodeModule provides AuthModule 

>     AuthModule

Reviewer: delete

  }
```

#### AuthModule

```go
type AuthModule interface {
    // SetModuleAuth makes the given module endpoints accessible only 
    // with authentication.
    SetModuleAuth(ctx context.Context, module string) error 
    // UnsetModuleAuth makes the given module endpoints accessible without
    // authentication.
    UnsetModuleAuth(ctx context.Context, module string) error
}

> ```

Reviewer: remove this

Reviewer 4: done

}
```

##### Metrics

```go
type MetricsModule interface {
 // TODO	
}

> ```

Author: Can you elaborate on this module pls? I don't really get its purpose and don't see how its API would look like

Author: So now it's much clear what it should be doing :)
